### PR TITLE
8318160: javac does not reject private method reference with type-variable receiver

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this PR which is fixing a long standing regression in javac. Basically javac is accepting code like:

```
import java.util.function.*;

class Test {
    private String asString() {
        return "bar";
    }

    static <T extends Test> Function<T, String> foo() {
        return T::asString;
    }
}
```

this code should be rejected on the grounds that ::asString is a private method in class `Test` and thus not a member of type variable `T`

TIA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8318203](https://bugs.openjdk.org/browse/JDK-8318203) to be approved

### Issues
 * [JDK-8318160](https://bugs.openjdk.org/browse/JDK-8318160): javac does not reject private method reference with type-variable receiver (**Bug** - P3)
 * [JDK-8318203](https://bugs.openjdk.org/browse/JDK-8318203): javac does not reject private method reference with type-variable receiver (**CSR**)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16210/head:pull/16210` \
`$ git checkout pull/16210`

Update a local copy of the PR: \
`$ git checkout pull/16210` \
`$ git pull https://git.openjdk.org/jdk.git pull/16210/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16210`

View PR using the GUI difftool: \
`$ git pr show -t 16210`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16210.diff">https://git.openjdk.org/jdk/pull/16210.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16210#issuecomment-1765473836)

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java [hunk 1]
```diff
@@ -3552,7 +3552,7 @@ class MethodReferenceLookupHelper extends ReferenceLookupHelper {
 
         MethodReferenceLookupHelper(JCMemberReference referenceTree, Name name, Type site,
                 List<Type> argtypes, List<Type> typeargtypes, MethodResolutionPhase maxPhase) {
-            super(referenceTree, name, types.skipTypeVars(site, true), argtypes, typeargtypes, maxPhase);
+            super(referenceTree, name, types.skipTypeVars(site, true, true), argtypes, typeargtypes, maxPhase);
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:3555]: I believe the right thing here should be to skip captured type variables but not other type variables

Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java:3555]: Not sure... if you have:
```
class Foo<X> {
   X get()
}
```
And then:

```
static Function<?, String> m2(Foo<? extends Test> foo) {
        return foo.get()::asString2;
    }
 ```
 
 Don't you end up with a similar issue?

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: I made several experiments and decided for the last commit pushed. We will need an additional field in class `MethodResolutionContext` basically pointing to the current `LookupHelper` which can now decide if a symbol is inherited at a given site or not. Semantically there are not changes for any other method the only semantic change will be when we are looking for the method in a method reference.

Comment (mcimadamore) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: I looked at the Resolve code some more, and also looked at the fix which likely caused this: 
https://bugs.openjdk.org/browse/JDK-8073842

So, `Resolve::findMethod` basically must NOT be called on a `site` Type that is a type variable. The original method resolution code enforces this in Attr::selectSym:

https://github.com/openjdk/jdk/blob/6fc35142315f1616fa35e415005c9483939c6920/src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java#L4575

Note that here:

1. If we see that the qualifier is a type variable, we recurse on the upper bound
2. If we find that we ended up accessing a symbol that was private, replace the symbol with an access error
3. call `accessBase` on the final symbol

I think a similar fix should be possible for method references. That is, as we do now, normalize the `site` using `skipTypeVars`, but also save the original `site` somewhere (as we already do). Note that the LookupHelper class has an `access` method, which is used to validate the result of the lookup (e.g. typically call `accessBase` on it). Effectively, this is a post-resolution step, pretty much like the one in `Attr`, so we can add whatever logic we want there (I think replicating the code in `Attr`  - see steps above - would be the best).

So, I don't think we need to add the new `isInherited` method - just overriding `access` to do the same thing we do in `Attr::selectSym` should be enough (I hope)

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Resolve.java]: I gave it another try and updated the code. One of my experiments was on this line not exactly what you are proposing and I was not vary happy with the end result. I think now it should be good, thanks

**Review by mcimadamore — APPROVED** [review]
Looks very good, thanks!

**Comment by vicente-romero-oracle** [comment]
> Looks very good, thanks!

thanks for the review! could you please review the CSR too?

**Pull request closed without merging**